Fixed #20869 -- Made CSRF tokens change every request.

shaib opened the pull request:
@PaulMcMillan would appreciate your review

raphaelm commented:
Hello Shai, my PR we talked about is #5600

shaib commented:
Reviewers, please consider the following points:

1) In this code I used the term `CSRF_KEY` for the hidden nonce which is padded and encrypted to create the csrf token. In view of the coming #5600 with its use of `CSRF_SESSION_KEY`, should we use a different name here, like `CSRF_NONCE` or `CSRF_SECRET`?

2) This was written to be added to master, but since it has security implications, should we backport it?

Thanks in advance.

funkybob commented on django/middleware/csrf.py:
```diff
@@ -27,6 +27,8 @@
 REASON_INSECURE_REFERER = "Referer checking failed - Referer is insecure while host is secure."
 
 CSRF_KEY_LENGTH = 32
+CSRF_TOKEN_LENGTH = CSRF_KEY_LENGTH + CSRF_KEY_LENGTH
```
Why not just "* 2" ...?

shaib replied:
I thought it was clearer that it's key-length + pad-length this way, but I may be wrong.

funkybob replied:
I realised after the comment that this may be the reason.

funkybob commented on django/middleware/csrf.py:
```diff
@@ -63,21 +101,28 @@ def rotate_token(request):
 def _sanitize_token(token):
     # Allow only alphanum
-    if len(token) > CSRF_KEY_LENGTH:
-        return _get_new_csrf_key()
```
Why has it changed from "on invalid length, return new token" to "on invalid length, return None"?

shaib replied:
I felt a little uneasy about the original mixing concerns; once it was clear that we need to sanitize both tokens (cookie and post/header), not just one, it became clear that we need to separate new-token-generation from sanitization.

apollo13 replied:
Looking at the code changes below in the middleware, I also see no clear reason for not returning a new token here to stay a little bit more compatible to with what we had before.

shaib replied:
Accepted. The separation of concerns is still correct IMO, but not crucial, and my "became clear" comment above refers to, essentially, optimizing an error case.

shaib commented:
Rebased, did TODOs, fixed python3 issues in tests, renamed the constant nonce from `CSRF_KEY` to `CSRF_NONCE` to be more compatible with #5600, fixed flake8 issues.

shaib commented:
isort is wrong, it ignores a leading `_` in an imported name but in truth `_` comes after all letters.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -63,21 +101,28 @@ def rotate_token(request):
-        return _get_new_csrf_key()
+    if len(token) != CSRF_TOKEN_LENGTH:
+        return None
+    if re.search('[^a-zA-Z0-9]', force_text(token)):
```
Wouldn't `isalnum()` be enough?

shaib replied:
I have always thought `_` is considered alphanumeric, which is wrong, but still, in Python 3 "alphanumeric" is in Unicode:
```python
In [4]: 'א'.isalnum()
Out[4]: True
```
So, no.

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -26,7 +26,9 @@
-CSRF_KEY_LENGTH = 32
+CSRF_NONCE_LENGTH = 32
+CSRF_TOKEN_LENGTH = 2 * CSRF_NONCE_LENGTH
+CSRF_ALLOWED_CHARS = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
```
`string.letters + string.digits`

shaib replied:
string.letters doesn't exist in Python 3 nor in six, so this required an enhancement to our own six, but, done.

apollo13 commented on docs/releases/1.10.txt:
```diff
@@ -325,6 +329,9 @@ Miscellaneous
 
 * ``utils.version.get_version()`` returns :pep:`440` compliant release
   candidate versions (e.g. '1.10rc1' instead of '1.10c1').
+* CSRF token values are now required to be strings of 64 alphanumerics; other
```
62 not 64

shaib replied:
NONCE_LENGTH is 32...

apollo13 replied:
Ah right, mixed it with `ALLOWED_CHARS`

collinanderson commented on docs/ref/csrf.txt:
```diff
@@ -234,14 +234,20 @@ The CSRF protection is based on the following things:
 
+   In order to protect against `BREACH`_ type attacks, the token is
+   not simply the nonce; a random pad is added to the nonce and used
+   to "encrypt" it. The pad is regenerated on every call to ``get_token()``,
```
We might want to avoid scare quotes. Maybe scramble?

funkybob commented on django/utils/six.py:
```diff
@@ -873,6 +873,7 @@ def python_2_unicode_compatible(klass):
 if PY3:
     memoryview = memoryview
     buffer_types = (bytes, bytearray, memoryview)
+    from string import ascii_letters
```
Not going to also go with uppercase/lowercase, whilst you're here?

timgraham replied:
I think this isn't needed. `from string import ascii_letters` works on Python 2 & 3.

shaib replied:
removed

timgraham commented on tests/csrf_tests/tests.py:
```diff
@@ -2,15 +2,21 @@
 from django.http import HttpRequest, HttpResponse
-from django.middleware.csrf import (
-    CSRF_KEY_LENGTH, CsrfViewMiddleware, get_token,
+from django.middleware.csrf import (                    # NOQA isort:skip
```
why disabling isort?

timgraham commented:
Should we revise the note in `docs/ref/middleware.txt` about BREACH? "Security researchers recently revealed..."

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -142,6 +142,12 @@ CSRF
+* The CSRF protection scheme was changed so that the form token value is changed on every
+  request (while keeping an invariant nonce which can be used to validate them).
+
+  To make this easier, we now force CSRF cookie values to be strings of
```
okay to remove this second paragraph? (seems redundant with the note below)

timgraham commented:
Edits: http://dpaste.com/120JES9
Please check the test failure too.

PaulMcMillan commented:
I'm not comfortable with this approach as-is. I'd like to propose (and do) an alternative overhaul of the CSRF protection, separate out BREACH protection, and avoid merging this in the meantime.

Objections to closing this and putting a different approach on the drawing board? Sorry I didn't chime in much sooner.

shaib commented:
@PaulMcMillan Can you elaborate a little about what makes you uncomfortable here? A new approach is certainly welcome as long as no changes have been committed, but I think it would be odd to close this one in its favor before it has even been presented.

aaugustin commented:
@PaulMcMillan Do you see drawbacks to merging this? Would it introduced new security risks or make further changes more difficult?

Django has a long tradition of committers wanting to do something, but not finding time. Sometimes better is the ennemy of good ;-)

apollo13 commented:
I think our last conclusion was that this change is not backwards
compatible for not much gain (BREACH is usually mitigated by not
compressing the output at all).

On Sat, Feb 27, 2016, 21:56 Aymeric Augustin <notifications@github.com>
wrote:

> @PaulMcMillan <https://github.com/PaulMcMillan> Do you see drawbacks to
> merging this? Would it introduced new security risks or make further
> changes more difficult?
>
> Django has a long tradition of committers wanting to do something, but not
> finding time. Sometimes better is the ennemy of good ;-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/5605#issuecomment-189723925>.
>

shaib commented:
@apollo13 I don't remember this being a consensus... There were problems with the approach that I set out to solve, then I found that some deeper work needed to be done because the current code is not doing what it looks like it's doing (and so, I think it's not doing what its authors intended) but what it is doing appears to be the right thing.

Backwards-compatibility issues exist in the latest version I made, but I don't think they're insurmountable. One way of mitigation is avoiding compression, but it seems many people are also using django-debreach -- and they shouldn't need to.

apollo13 commented:
On Sat, Feb 27, 2016 at 11:58 PM Shai Berger <notifications@github.com>
wrote:

> @apollo13 <https://github.com/apollo13> I don't remember this being a
> consensus...
>
Absolutely not, that was just the latest discussion I remember with Collin
and you on IRC.

Backwards-compatibility issues exist in the latest version I made, but I
> don't think they're insurmountable.


No argument there, I didn't mean "latest coclusion" to be authoritative in
any way -- just that it was my last status of when we talked about it
(sorry if my words were chosen poorly)


> One way of mitigation is avoiding compression, but it seems many people
> are also using django-debreach -- and they shouldn't need to.


I just hope they do not think that this makes them safe from BREACH like
attacks.

shaib commented:
buildbot, test this please.

shaib commented:
Doc build error fixed, patch is now ready for final reviews

timgraham commented:
Shai, did you notice the test failure on Python 3?

shaib commented:
No, I hadn't. Will take a look. Thanks for calling my attention to it.

shaib commented:
Well, of course, it was my incorrect understanding of `warnings.catch_warnings()`. Should be fixed now.

shaib commented:
buildbot, test on windows.

shaib commented:
@apollo13 @PaulMcMillan  @collinanderson this is supposed to be fully backwards compatible; doesn't change cookies when it doesn't need to (there are tests for that); made the code in `process_response()` clearer by not using the same attribute name (`csrf_processing_done`) for both request and response. I would appreciate a final review, especially details about Paul's discomfort, but this has already been through some review with concerns addressed.

timgraham commented on django/middleware/csrf.py:
```diff
@@ -63,19 +99,35 @@ def rotate_token(request):
+    elif len(token) == CSRF_TOKEN_LENGTH:
+        return token
+    elif len(token) == CSRF_NONCE_LENGTH:
+        # For backwards compatibility, accept such values as unpadded nonces.
```
Is this branch a candidate for removal in some future release? Either way, would be helpful to elaborate on how the backwards-compatibility case arises. I think it's upgrading Django in the presence of existing token set by older versions of Django, but would like t to confirm.

collinanderson replied:
I personally generate the NONCE's on the client side using javascript (so I don't need the token in the response body). It would be nice to continue long term to accept the raw cookie value as a valid csrf token.

collinanderson replied:
It would allow our "AJAX" method to continue working unchanged:
https://docs.djangoproject.com/en/dev/ref/csrf/#ajax

shaib replied:
The cookie value is also padded and ciphered, and so it is a valid token and the existing method continues to work.

collinanderson replied:
Ahh. Got it. That makes sense. (Seems fine to me if we eventually remove it. Maybe raise a deprecation warning.)

shaib replied:
@timgraham fixed

timgraham commented on django/middleware/csrf.py:
```diff
@@ -63,19 +99,35 @@ def rotate_token(request):
+def _compare_padded_tokens(request_csrf_token, csrf_token):
+    if not(request_csrf_token and csrf_token):
+        return False
+    # Assume both arguments to be sanitized -- that is, either strings of
```
to be -> are

shaib replied:
fixed

timgraham commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,36 @@ The CSRF protection is based on the following things:
+   In order to protect against `BREACH`_ attacks, the token is not simply the
+   nonce; a random pad is added to the nonce and used to scramble it.
+
+.. _BREACH: http://breachattack.com/
```
please move this line to the bottom of the section -- as it is now it causes incorrect rendering.

shaib replied:
Didn't notice this before (and this is not a new part of the PR... maybe Sphinx changed?), thanks. Fixed.

timgraham commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,36 @@ The CSRF protection is based on the following things:
-   cookie.
+   outgoing POST forms. The value of this field is, again, the value of the
+   nonce, with a pad which is both added to it and used to scramble it. The pad
+   is regenerated on every call to ``get_token()``, so that the form field
```
no comma needed

shaib replied:
removed

timgraham commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,36 @@ The CSRF protection is based on the following things:
    TRACE, a CSRF cookie must be present, and the 'csrfmiddlewaretoken' field
    must be present and correct. If it isn't, the user will get a 403 error.
 
+   When validating the 'csrfmiddlewaretoken' field value, it is only the nonce,
```
chop "it is" and "which" in the next line

shaib replied:
chopped

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -212,6 +212,16 @@ CSRF
   accepts an optional ``template_name`` parameter, defaulting to
   ``'403_csrf.html'``, to control the template used to render the page.
 
+* To protect against `BREACH`_ attacks, the CSRF protection scheme is changed
```
the CSRF protection changes the form token on every request

shaib replied:
simplified along these lines ("the CSRF protection mechanism now  changes the form token")

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -212,6 +212,16 @@ CSRF
+  so that the form token value is changed on every request (while keeping an
+  invariant nonce which can be used to validate them).
+
+  To make this easier, the CSRF cookie values are forced to be strings of
```
Probably don't need this bit here.

shaib replied:
I think we do. It's a user-visible (even end-user-visible) change.

shaib replied:
Oh, you mean because it is included in the backward-incompat part. Agreed.

timgraham commented on docs/releases/1.10.txt:
```diff
@@ -657,6 +667,12 @@ Miscellaneous
+  of 32 alphanumerics, as set by older versions of Django by default, are
+  automatically replaced by strings of 64 characters. Other values are
+  considered invalid. This should only affect developers or users who replace
+  these tokens explicitly.
```
chop "explicitly"?

shaib replied:
chopped.

timgraham commented on tests/csrf_tests/tests.py:
```diff
@@ -295,6 +347,35 @@ def test_token_node_with_new_csrf_cookie(self):
 
+    def test_cookie_not_reset_on_accepted_request(self):
+        """
+        The csrf token used in posts is changed on every request (although stays equivalent).
```
wrap docstrings at 79 characters

shaib replied:
Wrapped. Wasn't aware of this rule; where does it come from? It is not warned on by flake8.

timgraham replied:
I don't think flake8 does any checking of docstrings. The 79 char suggested limit is mentioned in https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style.

timgraham commented on tests/csrf_tests/tests.py:
```diff
@@ -295,6 +347,35 @@ def test_token_node_with_new_csrf_cookie(self):
+        resp = CsrfViewMiddleware().process_response(req, resp)
+        csrf_cookie = resp.cookies.get(settings.CSRF_COOKIE_NAME, None)
+        if csrf_cookie:
+            self.assertEqual(csrf_cookie.value, self._csrf_id_cookie,
```
use hanging indent please.

shaib replied:
Not sure what you mean; if you think line 362 needs to be indented more than 363, I disagree. If you mean something else, please elaborate.

timgraham replied:
I suggest:
```
self.assertEqual(
    csrf_cookie.value, self._csrf_id_cookie,
    '.....'
)
```

shaib replied:
Accepted

shaib commented:
@timgraham thanks for all these additional comments. I will take care of them this week.

timgraham commented on django/middleware/csrf.py:
```diff
@@ -63,19 +99,37 @@ def rotate_token(request):
+        return token
+    elif len(token) == CSRF_NONCE_LENGTH:
+        # Older Django versions set cookies to values of CSRF_NONCE_LENGTH
+        # alphanumeric characters. For backwards compatibility,  we accept
```
Chop "  we"

shaib replied:
chopped

timgraham commented:
@PaulMcMillan @apollo13 @dstufft  -- any objections to merging this by Monday for 1.10 alpha?

apollo13 commented on django/middleware/csrf.py:
```diff
@@ -63,19 +99,37 @@ def rotate_token(request):
+
+
+def _compare_padded_tokens(request_csrf_token, csrf_token):
+    if not(request_csrf_token and csrf_token):
```
Is that check actually needed? This seems to be only needed if a token is `None` -- which would not happen if we return a token from `sanitize_token` instead of None :D

shaib replied:
Removed as part of accepting the above comments about `_sanitize_token()`

apollo13 commented:
Given that there is no reaction from @PaulMcMillan I am somewhat +0 on merging this patch. I have one issue with it though. What this patch refers to as `csrf_nonce` (ie the one unpadded from `request.META['CSRF_COOKIE']`) is not actually a nonce but the real (unencrypted) token. A nonce (cryptographically speaking) is something only used once (ie what this patch refers to as `pad`). The token does not change, it stays constant over multiple requests. Naming it nonce breaks quite a few assumptions (and made it harder for me to understand this patch, since I immediatelly associated nonce with the "random pad" and not the token). Since this is somewhat security sensitive, I think a nonce should do what one would usually expect from a nonce.

So I would like to so the following rewrite before merging:
 * the constant part should be named `key` (ie in similarity to what we had before)
 * keep `pad` or rename it to `nonce`, since this is the actual changing part
 * the token would then be `nonce + cipher(key, nonce)`
Adjust function names accordingly where appropriate.

Given those changes and approval from @shaib and no objections from @PaulMcMillan or @dstufft I think we should merge this for 1.10 -- to the best of my knowledge I cannot find any more flaws in this patch (that said, I am somewhat at the end of my comfort zone with regard to security).

MarkusH commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,34 @@ The CSRF protection is based on the following things:
 1. A CSRF cookie that is set to a random value (a session independent nonce, as
```
"nonce" is technically not correct, I think. At least for the previous CSRF implementation as it was used more than once.

shaib replied:
See my  other comment about correctness, but the tern was changed anyway to avoid confusion.

MarkusH commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,34 @@ The CSRF protection is based on the following things:
-   but since there is no way to set a cookie that never expires, it is sent with
-   every response that has called ``django.middleware.csrf.get_token()``
-   (the function used internally to retrieve the CSRF token).
+   This cookie is set by ``CsrfViewMiddleware``.  It is sent with every
```
Double space after the `CsrfViewMiddleware`

shaib replied:
fixed

MarkusH commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,34 @@ The CSRF protection is based on the following things:
+   set on the request.
+
+   In order to protect against `BREACH`_ attacks, the token is not simply the
+   nonce; a random pad is added to the nonce and used to scramble it.
```
Added where? Front? Back? Middle?

shaib replied:
s/added/prepended/

MarkusH commented on docs/ref/csrf.txt:
```diff
@@ -222,24 +222,34 @@ The CSRF protection is based on the following things:
+   When validating the 'csrfmiddlewaretoken' field value, only the nonce,
+   not the full token, is compared with the nonce in the cookie value.
+   This allows the use of ever-changing tokens. While each request may use its
+   own token, the nonce remains common to all.
```
I think you mixed up "token" and "nonce" here, as pointed out by @apollo13 earlier.

shaib replied:
replacing "nonce" with "secret" clears it up, I hope.

MarkusH commented on tests/csrf_tests/tests.py:
```diff
@@ -2,15 +2,21 @@
+import re
+import warnings
+
+from six import text_type
```
Do you mean `django.utils.six` ?

shaib replied:
Of course.

shaib commented:
Thanks for this; I will do my best to fix according to your comments in time, 
but I would just like to clarify the point of the "nonce" terminology: I am 
not a security expert and not aware of the terms used in security literature, 
and my use of the term "nonce" is simply consistent with the way it has been 
used before. In particular, you can find the phrase "session independent nonce" 
(which strongly implies more-than-once use) in current Django documentation.

So just to make sure -- Django documentation has been wrong all these years? 
Can you point some reference to this?

apollo13 commented:
@shaib Yes, I think the word in the documentation is unfortunate, the code always referred to it as key/token IIRC. I am obviously also no security expert, it was just something which felt weird so I looked it up (sorry for wikipedia links following):

apollo13 commented:
@shaib Yes, I think the word in the documentation is unfortunate, the code always referred to it as key/token IIRC. I am obviously also no security expert, it was just something which felt weird when reviewing again (and probably was also the reason why I felt so uneasy when I looked at the code last time). https://en.wikipedia.org/wiki/Cryptographic_nonce provides imo a nice explanation.

jdufresne commented on django/middleware/csrf.py:
```diff
@@ -36,8 +39,38 @@ def _get_failure_view():
+    """
+    pad = _get_new_csrf_nonce()
+    chars = CSRF_ALLOWED_CHARS
+    pairs = zip([chars.index(x) for x in nonce], [chars.index(x) for x in pad])
```
I think this could use a generator expression instead of list comprehension to avoid additional temporary lists in memory. For example:

```
pairs = zip((chars.index(x) for x in nonce), (chars.index(x) for x in pad))
```

(Same in `_unpad_cipher_token()`)

shaib replied:
Accepted (also imported `zip` from `django.utils.six.moves` to make sure we use the iterator version on Python 2 as well)

shaib commented:
FWIW, it appears that a "session independent nonce" is indeed a thing: http://www-inst.cs.berkeley.edu/~cs161/sp16/slides/2.10.SessionCSRF.pdf page 42. Your second link also explains that, in general use, "nonce" can refer to "a {number, word} used for one **purpose**", not just "used **once**".

I am changing the terminology, but I don't want to use the term "key" as that is way overloaded; I'm using "secret" or "csrf_secret" for the constant part, "token" for texts that go over the wire, "salt" (as @apollo13 suggested) for the part that gets added.

The function `_get_new_csrf_nonce()` was renamed to `_get_new_csrf_string()` as it is used to produce both secrets and salts.

shaib commented:
I added the fixes in separate commits to make reviewing easier; if all looks well, I'll squash them and push to master.

timgraham commented on django/middleware/csrf.py:
```diff
@@ -106,27 +108,25 @@ def rotate_token(request):
 
 def _sanitize_token(token):
-    # Allow only alphanum
+    # Allow only alphanumerics from ASCII
```
I'd say "ASCII alphanumerics"

timgraham commented on docs/ref/csrf.txt:
```diff
@@ -218,23 +218,23 @@ How it works
 
-1. A CSRF cookie that is set to a random value (a session independent nonce, as
-   it is called), which other sites will not have access to.
+1. A CSRF cookie that is based on a random secret value, which other
```
"sites" fits on this line

timgraham commented on docs/ref/csrf.txt:
```diff
@@ -218,23 +218,23 @@ How it works
 2. A hidden form field with the name 'csrfmiddlewaretoken' present in all
    outgoing POST forms. The value of this field is, again, the value of the
-   nonce, with a pad which is both added to it and used to scramble it. The pad
+   secret, with a salt which is both added to it and used to scramble it. The salt
```
line > 80 chars
(could chop "both" or reflow)

timgraham commented on docs/ref/middleware.txt:
```diff
@@ -147,6 +146,13 @@ It will NOT compress content if any of the following are true:
 
+.. versionchanged:: 1.10
+
+    In releases before 1.10, Django's own CSRF protection mechanism was
```
In releases before 1.10, -> In older versions,
chop "own"

timgraham commented on tests/template_backends/test_dummy.py:
```diff
@@ -2,9 +2,13 @@
 from django.http import HttpRequest
-from django.middleware.csrf import CsrfViewMiddleware, get_token
+from django.middleware.csrf import (
+    CsrfViewMiddleware, _compare_padded_tokens as equivalent_tokens, get_token,
```
I guess it should be `_compare_salted_tokens` (rather than "padded") to fix build errors.

shaib commented:
Fixed all issues raised by @timgraham and squashed.
